plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1363 @@
+        """
+        Return a `dict` containing `~plasmapy_nei.eigen.EigenData` instances
+        for each element.
+        """
+        return self._EigenDataDict
+
+    @property
+    def initial(self) -> IonizationStates:
+        """
+        Return the ionization states of the plasma at the beginning of
+        the simulation.
+        """
+        return self._initial
+
+    @initial.setter
+    def initial(self, initial_states: IonizationStates):
+        if isinstance(initial_states, IonizationStates):
+            self._initial = initial_states
+            self._elements = (
+                initial_states.ionic_fractions.keys()
+            )  # TODO IonizationStates
+        elif initial_states is None:
+            self._ionstates = None
+        else:
+            raise TypeError("Expecting an IonizationStates instance.")
+
+    @property
+    def results(self) -> SimulationResults:
+        """
+        Return the `~plasmapy_nei.nei.SimulationResults` class instance that
+        corresponds to the simulation results.
+
+        """
+        if self._results is not None:
+            return self._results
+        else:
+            raise AttributeError("The simulation has not yet been performed.")
+
+    @property
+    def final(self) -> IonizationStates:
+        """
+        Return the ionization states of the plasma at the end of the
+        simulation.
+        """
+        try:
+            return self._final
+        except AttributeError:
+            raise NEIError("The simulation has not yet been performed.") from None
+
+    def _initialize_simulation(self):
+
+        self._results = SimulationResults(
+            initial=self.initial,
+            n_init=self.hydrogen_number_density(self.time_start),
+            T_e_init=self.electron_temperature(self.time_start),
+            max_steps=self.max_steps,
+            time_start=self.time_start,
+        )
+        self._old_time = self.time_start.to(u.s)
+        self._new_time = self.time_start.to(u.s)
+
+    def simulate(self) -> SimulationResults:
+        """
+        Perform a non-equilibrium ionization simulation.
+
+        Returns
+        -------
+        results: `~plasmapy_nei.classes.Simulation`
+            The results from the simulation (which are also stored in
+            the ``results`` attribute of the `~plasmapy_nei.nei.NEI`
+            instance this method was called from.
+
+        """
+
+        self._initialize_simulation()
+
+        for step in range(self.max_steps):
+            try:
+                self.set_timestep()
+                self.time_advance()
+            except StopIteration:
+                break
+            except Exception as exc:
+                raise NEIError(f"Unable to complete simulation.") from exc
+
+        self._finalize_simulation()
+
+        # Is there a way to use the inspect package or something similar
+        # to only return self.results if it is in an expression where
```

Comment:
What did you have in mind?

Reply:
I wouldn't have `simulate()` return results at all, since that encourages users to separate the results from the originating simulation.   I would encourage  the use of...

```
>>> sim.simulate()
>>> sim.results.plot_something()
```
Or, if the result was a `xarray.Dataset` then one could do...

```
>>> sim.simulate()
>>> sim.results.T_e.plot()
```